Improve render docs

eteq opened the pull request:
This PR makes some changes to the docstring of the `Model.render` method.  I found the current in-master docstring to be a bit difficult to follow, so I'm hoping this will make things a bit clearer.

I'd appreciate if @patti and/or @cdeil could take a look at this and double-check I didn't say anything wrong (or if you think may change made it *less* clear I'm happy to here that and make changes appropriately!)

patti commented:
This all looks good to me :thumbsup:

embray commented:
I've been meaning to improve the docs for this too--the other day I had some people confused about exactly what this method does, and the docstring admittedly didn't help much in that case.

embray commented on astropy/modeling/core.py:
```diff
@@ -1074,22 +1074,28 @@ def evaluate(self, *args, **kwargs):
+            to the ``out`` array. It should have the property that
+            ``self(coords)`` yields the same shape as ``out``.  If ``out`` is
+            not specified, ``coords`` will be used to determine the shape of the
+            returned array.
```
I think this could still be clarified further--that by default the coordinate grid to evaluate the model on is determined by the bounding box.

astrofrog commented on astropy/modeling/core.py:
```diff
@@ -1074,22 +1074,28 @@ def evaluate(self, *args, **kwargs):
         """
-        Evaluates a model on an input array. Evaluation is limited to
-        a bounding box if the `Model.bounding_box` attribute is set.
+        Evaluates a model on an input array.  I.e., "discretizes" the model.
```
Just to clarify, it evaluates it at fixed positions, but doesn't evaluate it over the whole pixel like discretize_model in http://docs.astropy.org/en/stable/api/astropy.convolution.discretize_model.html#astropy.convolution.discretize_model right? I wonder if we need to make that clearer.

eteq replied:
Good point, @astrofrog.  This just came up in #4572, actually.  Will clear that up.

I wonder if we should consider adding a switch in `render` to allow it to do `discretize_model`, too?  I'm thinking the typical user (me included, in this case!) won't think to look at convolution for this.

nden commented:
I am still a bit confused about how `bounding_box` works.
I am giving a few examples below. I'd appreciate if anyone can clarify the questions there.

I'm using a 1D model in the example:
```
g2 = models.Gaussian1D(10, 6, 0.2)
print(g2.bounding_box)
(4.9000000000000004, 7.0999999999999996)
```
It seems like the inputs are not converted to float. The default `evaluate` method converts inputs (and parameters) to float, so this was a surprise.
```
x1 = np.arange(0, 7)
a1 = x1.copy()
yr1 = g2.render(a1)
print(yr1)
array([ 0,  1,  2,  3,  4,  5, 16])
```

The `bounding_box` adds the result to the input array instead of returning the evaluated function at those inputs. Is this the intention?

```
x1=np.arange(0, 7, dtype=np.float)
g2.render(a1)
array([  0. , 1., 2., 3., 4. , 5.00003727, 16.])
```

This example shows that negative inputs are evaluated outside the bounding box.
```
x = np.arange(-3, 6, dtype=np.float)
a = x.copy()
yr = g2.render(a)
print(yr)
[ -3., -2., -1., 0., 1., 2.00003727, 13., 4.00003727, 5.]
print(x)
[-3., -2., -1.,  0.,  1.,  2.,  3.,  4.,  5.]
```
Given the `bounding_box` above I'd expect only the last two elements of the input array to be evaluated. But perhaps I misunderstand this and the intention is not to filter out inputs which are outside the `bounding_box`.

patti commented:
@nden

Inputs are not converted to float. That's an oversight. I actually get a `TypeError` using 2.7 so this should definitely be fixed. 

The render method was designed to render or grid the model to an (n-dimensional) image. The input `arr` is the grid or image. The model is evaluated at (and defined wrt) the coordinates of the grid. So in your last example, the model is centered at the 6th element of `a` and the result is added to `a`.

I hope that helps. This certainly needs clarification in the docs. Unfortunately I don't have the time to do this myself right now (I'm defending my thesis in 2 weeks) but I'll do my best to get around to when I find some time.

nden commented:
Thanks @patti . This certainly helps.

eteq commented:
Alright, I addressed the comments above.  I also checked that the docs built locally locally, so in the name of expediency for release I'm just going to merge this without waiting on travis.
